### PR TITLE
Fix AdaptiveCapacityDictionary assert with 11 elements 

### LINKED ISSUE #34307: AdaptiveCapacityDictionary assert with 11 elements
The client sent 11 cookies, the threshold for ACD is 10.

While this is only an assert, I expect being in this invalid state will cause some weird side-effects, potentially discarding cookies when doing incorrect resizes.
```
Process terminated. Assertion failed.
at Microsoft.AspNetCore.Internal.AdaptiveCapacityDictionary`2.set_Item(TKey key, TValue value) in C:\Github\aspnetcore\src\Shared\Dictionary\AdaptiveCapacityDictionary.cs:line 141
at Microsoft.Net.Http.Headers.CookieHeaderParserShared.TryParseValues(StringValues values, IDictionary`2 store, Boolean enableCookieNameEncoding, Boolean supportsMultipleValues) in C:\Github\aspnetcore\src\Http\Shared\CookieHeaderParserShared.cs:line 37
at Microsoft.AspNetCore.Http.RequestCookieCollection.ParseInternal(StringValues values, Boolean enableCookieNameEncoding) in C:\Github\aspnetcore\src\Http\Http\src\Internal\RequestCookieCollection.cs:line 76
at Microsoft.AspNetCore.Http.RequestCookieCollection.Parse(StringValues values) in C:\Github\aspnetcore\src\Http\Http\src\Internal\RequestCookieCollection.cs:line 65
at Microsoft.AspNetCore.Http.Features.RequestCookiesFeature.get_Cookies() in C:\Github\aspnetcore\src\Http\Http\src\Features\RequestCookiesFeature.cs:line 74
at Microsoft.AspNetCore.Http.DefaultHttpRequest.get_Cookies() in C:\Github\aspnetcore\src\Http\Http\src\Internal\DefaultHttpRequest.cs:line 145
at Microsoft.AspNetCore.Authentication.Cookies.ChunkingCookieManager.GetRequestCookie(HttpContext context, String key) in C:\Github\aspnetcore\src\Shared\ChunkingCookieManager\ChunkingCookieManager.cs:line 101
at Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler.ReadCookieTicket() in C:\Github\aspnetcore\src\Security\Authentication\Cookies\src\CookieAuthenticationHandler.cs:line 143
at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in System.Private.CoreLib.dll:token 0x6004ea3+0x28
at Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler.ReadCookieTicket() in Microsoft.AspNetCore.Authentication.Cookies.dll:token 0x6000034+0x1f
at Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler.EnsureCookieTicket() in C:\Github\aspnetcore\src\Security\Authentication\Cookies\src\CookieAuthenticationHandler.cs:line 78
```
Stack walk:

_dictionaryStorage and _arrayStorage are not both supposed to have a value
https://github.com/dotnet/aspnetcore/blob/12c57a5e0ae5bc883e3ff46af90887b2965d5563/src/Shared/Dictionary/AdaptiveCapacityDictionary.cs#L579-L583
Indexer.Set:
https://github.com/dotnet/aspnetcore/blob/12c57a5e0ae5bc883e3ff46af90887b2965d5563/src/Shared/Dictionary/AdaptiveCapacityDictionary.cs#L141
https://github.com/dotnet/aspnetcore/blob/12c57a5e0ae5bc883e3ff46af90887b2965d5563/src/Http/Shared/CookieHeaderParserShared.cs#L37
Collection created:
https://github.com/dotnet/aspnetcore/blob/12c57a5e0ae5bc883e3ff46af90887b2965d5563/src/Http/Http/src/Internal/RequestCookieCollection.cs#L73-L76
(related: using StringValues.Count as the initial capacity is actually a bad idea, cookies are supposed to be sent in a single header so there should only be 0 or 1 entries in the StringValues. (except for a Kestrel HTTP/2 https://github.com/dotnet/aspnetcore/issues/26461 bug which I think is necessary to hit this issue).

Bug: The capacity constructor initializes both fields.
https://github.com/dotnet/aspnetcore/blob/12c57a5e0ae5bc883e3ff46af90887b2965d5563/src/Shared/Dictionary/AdaptiveCapacityDictionary.cs#L79-L80

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes #34307 

AdaptiveCapacityDictionary starts with an Array and then transitions to a Dictionary after 10 elements. It uses null checks on these fields to figure out which mode it's in. However, one constructor was initializing both fields and put the object in an invalid state.

This is a Debug.Assert failure that Hao noticed when manually testing something. It's a 6.0 regression introduced with the new AdaptiveCapacityDictionary. Neither of us saw any incorrect product behavior related to the bug, but there's potential for it given the invalid state the object is in.

Related: the cookie parser shouldn't be assuming that the StringValues.Count is the cookie count. The Cookie header is supposed to be a single value so StringValues.Count would normally be 0 or 1. There is a Kestrel Http/2 bug (#26461) where you can get more values, which is how Hao ran into this issue.

### CODE REVIEW

File: src/Http/Http/src/Internal/RequestCookieCollection.cs [hunk 1]
```diff
@@ -70,7 +70,9 @@ internal static RequestCookieCollection ParseInternal(StringValues values, bool
             {
                 return Empty;
             }
-            var collection = new RequestCookieCollection(values.Count);
+
+            // Do not set the collection capacity based on StringValues.Count, the Cookie header is supposed to be a single combine value.
```
Comment (BrennanConroy) [inline src/Http/Http/src/Internal/RequestCookieCollection.cs:74]: nit:
```suggestion
            // Do not set the collection capacity based on StringValues.Count, the Cookie header is supposed to be a single combined value.
```

File: src/Shared/Dictionary/AdaptiveCapacityDictionary.cs [hunk 2]
```diff
@@ -95,6 +94,7 @@ internal AdaptiveCapacityDictionary(IEnumerable<KeyValuePair<TKey, TValue>> valu
         {
             _comparer = comparer ?? EqualityComparer<TKey>.Default;
 
+            Debug.Assert(capacity <= DefaultArrayThreshold);
```
Comment (BrennanConroy) [inline src/Shared/Dictionary/AdaptiveCapacityDictionary.cs:97]: Why doesn't this ctor "fallback" to dictionary with a big enough size?

File: src/Http/Http/test/RequestCookiesCollectionTests.cs [hunk 3]
```diff
@@ -38,5 +38,13 @@ public void AppContextSwitchUnEscapesKeysAndValues(string input, string expected
             Assert.Equal(expectedKey, cookies.Keys.Single());
             Assert.Equal(expectedValue, cookies[expectedKey]);
         }
+
+        [Fact]
+        public void ParseManyCookies()
+        {
+            var cookies = RequestCookieCollection.Parse(new StringValues(new[] { "a=a", "b=b", "c=c", "d=d", "e=e", "f=f", "g=g", "h=h", "i=i", "j=j", "k=k", "l=l" }));
+
+            Assert.Equal(12, cookies.Count);
```
Comment (HaoK) [inline src/Http/Http/test/RequestCookiesCollectionTests.cs:47]: Was this failing before this change?

Comment (HaoK) [reply src/Http/Http/test/RequestCookiesCollectionTests.cs:47]: Maybe we should also add a crazy big variant of this that parses 1000 cookies as well?

File: src/Shared/Dictionary/AdaptiveCapacityDictionary.cs (hunk 2 above)
Comment (Tratcher) [reply src/Shared/Dictionary/AdaptiveCapacityDictionary.cs:97]: This constructor is only used in tests.

Comment (BrennanConroy) [reply src/Shared/Dictionary/AdaptiveCapacityDictionary.cs:97]: It should just call the other ctor, no reason to have split code for testing and non-testing

File: src/Http/Http/test/RequestCookiesCollectionTests.cs (hunk 3 above)
Comment (Tratcher) [reply src/Http/Http/test/RequestCookiesCollectionTests.cs:47]: This test did trigger the Debug.Assert inside Parse. The Count Assert was just to make sure items weren't being lost.

Comment (Tratcher) [reply src/Http/Http/test/RequestCookiesCollectionTests.cs:47]: > Maybe we should also add a crazy big variant of this that parses 1000 cookies as well?

10 is the only boundary condition in todays logic, anything bigger than that ends up testing Dictionary.

**Review by HaoK — APPROVED** [review]

File: src/Shared/Dictionary/AdaptiveCapacityDictionary.cs (hunk 2 above)
Comment (Tratcher) [reply src/Shared/Dictionary/AdaptiveCapacityDictionary.cs:97]: Let's see if I can just delete it, the test usage doesn't look novel.

**Review by BrennanConroy — APPROVED** [review]

**Comment by davidfowl** [comment]
Nice find!

**Pull request merged**